core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -376,12 +381,36 @@ class Partition(val topic: String,
   }
 
   /**
+   * Check and maybe increment the low watermark of the partition;
+   * this function can be triggered when any replica's logStartOffset changed
```

Comment:
BTW, the reason `maybeIncrementLeaderLW()` is not called when the old segments are deleted is that the concept of `LW` and partition is at a higher level than `LogManager`, `LogCleaner` and `Log`. Ideally those classes such as `LogManager` do not contain reference to the higher level class such as `Partition`.